jps-plugin/src/org/intellij/erlang/jps/builder/ErlangDirtyFilesListsBuilder.java

```diff
@@ -0,0 +1,257 @@
+                                     @NotNull GraphGenerator<String> graph,
+                                     @NotNull Set<String> dirtyFiles) {
+    if (dirtyFiles.contains(filePath)) return;
+    dirtyFiles.add(filePath);
+    Iterator<String> childIterator = graph.getOut(filePath);
+    while (childIterator.hasNext()) {
+      findDirtyFiles(childIterator.next(), graph, dirtyFiles);
+    }
+  }
+
+  @NotNull
+  private static ErlangModuleBuildOrder getOrCreateModuleBuildOrder(@NotNull String moduleName,
+                                                                    @NotNull Map<String, ErlangModuleBuildOrder> buildOrders) {
+    ErlangModuleBuildOrder erlangModuleBuildOrder = buildOrders.get(moduleName);
+    if (erlangModuleBuildOrder != null) {
+      return erlangModuleBuildOrder;
+    }
+    buildOrders.put(moduleName, new ErlangModuleBuildOrder());
+    return buildOrders.get(moduleName);
+  }
+
+  private static class SortedModuleDependencyGraph implements GraphGenerator.SemiGraph<String> {
+    private final LinkedHashMap<String, List<String>> myNodePathsMap = ContainerUtil.newLinkedHashMap();
+
+    public SortedModuleDependencyGraph(@NotNull ErlangProjectBuildOrder dependencyTree) {
+      for (ErlangFileDescriptor node : dependencyTree.myNodes) {
+        myNodePathsMap.put(node.myPath, node.myDependencies);
+      }
+    }
+
+    @Override
+    public Collection<String> getNodes() {
+      return myNodePathsMap.keySet();
+    }
+
+    @Override
+    public Iterator<String> getIn(String node) {
+      return myNodePathsMap.get(node).iterator();
+    }
+  }
+
+  private static class BuildOrdersCreator {
+    private final Map<String, SourceRootInfo> mySourceInfoIndex;
+    private final List<ErlangFileDescriptor> mySortedNodes;
+
+    private BuildOrdersCreator(@NotNull JpsProject project,
+                               @NotNull List<ErlangFileDescriptor> sortedNodes) {
+      mySortedNodes = sortedNodes;
+      mySourceInfoIndex = ContainerUtil.newHashMap();
+      for (JpsModule myModule : project.getModules()) {
+        for (JpsTypedModuleSourceRoot<JavaSourceRootProperties> root : myModule.getSourceRoots(JavaSourceRootType.SOURCE)) {
+          mySourceInfoIndex.put(root.getFile().getAbsolutePath(), new SourceRootInfo(myModule.getName(), false));
+        }
+        for (JpsTypedModuleSourceRoot<JavaSourceRootProperties> root : myModule.getSourceRoots(JavaSourceRootType.TEST_SOURCE)) {
+          mySourceInfoIndex.put(root.getFile().getAbsolutePath(), new SourceRootInfo(myModule.getName(), true));
+        }
+      }
+    }
+
+    @NotNull
+    public static Map<String, ErlangModuleBuildOrder> createBuildOrders(@NotNull JpsProject project,
+                                                                        @NotNull List<ErlangFileDescriptor> nodes,
+                                                                        @NotNull Set<String> dirtySortedErlangFiles) {
```

Comment:
Are you really read code or only method declaration? What about `getSortedDirtyModules` method?) Name `dirtySortedErlangFiles` is not exact,perhaps this is the problem. I'm rename that argument.